src/Data/Argonaut/Decode/Implementation.purs

```diff
@@ -0,0 +1,163 @@
+import Data.Array.NonEmpty as NEA
+import Data.Bifunctor (lmap, rmap)
+import Data.Either (Either(..), note)
+import Data.Identity (Identity(..))
+import Data.Int (fromNumber)
+import Data.List (List, fromFoldable)
+import Data.List as L
+import Data.List.NonEmpty (NonEmptyList)
+import Data.List.NonEmpty as NEL
+import Data.Map as M
+import Data.Maybe (maybe, Maybe(..))
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.Set as S
+import Data.String (CodePoint, codePointAt)
+import Data.Traversable (traverse)
+import Data.TraversableWithIndex (traverseWithIndex)
+import Data.Tuple (Tuple(..))
+import Foreign.Object as FO
+
+type Decoder a = Json -> Either String a
+
+decodeIdentity :: ∀ a . Decoder a -> Decoder (Identity a)
+decodeIdentity decoder j = Identity <$> decoder j
+
+decodeMaybe :: ∀ a . Decoder a -> Decoder (Maybe a)
+decodeMaybe decoder j
+  | isNull j = pure Nothing
+  | otherwise = Just <$> decoder j
+
+decodeTuple :: ∀ a b . Decoder a -> Decoder b -> Decoder (Tuple a b)
+decodeTuple decoderA decoderB j = decodeArray Right j >>= f
+  where
+  f :: Array Json -> Either String (Tuple a b)
+  f [a, b] = Tuple <$> decoderA a <*> decoderB b
+  f _ = Left "Couldn't decode Tuple"
+
+decodeEither :: ∀ a b . Decoder a -> Decoder b -> Decoder (Either a b)
+decodeEither decoderA decoderB j =
+  lmap ("Couldn't decode Either: " <> _) $
+    decodeJObject j >>= \obj -> do
+      tag <- maybe (Left "Expected field 'tag'") Right $ FO.lookup "tag" obj
+      val <- maybe (Left "Expected field 'value'") Right $ FO.lookup "value" obj
+      case toString tag of
+        Just "Right" -> Right <$> decoderB val
+        Just "Left" -> Left <$> decoderA val
+        _ -> Left "'tag' field was not \"Left\" or \"Right\""
+
+decodeNull :: Decoder Unit
+decodeNull = caseJsonNull (Left "Value is not a null") (const $ Right unit)
+
+decodeBoolean :: Decoder Boolean
+decodeBoolean = caseJsonBoolean (Left "Value is not a Boolean") Right
+
+decodeNumber :: Decoder Number
+decodeNumber = caseJsonNumber (Left "Value is not a Number") Right
+
+decodeInt :: Decoder Int
+decodeInt =
+  maybe (Left "Value is not an Integer") Right
+    <<< fromNumber
+    <=< decodeNumber
+
+decodeString :: Decoder String
+decodeString = caseJsonString (Left "Value is not a String") Right
+
+decodeNonEmpty_Array :: ∀ a . Decoder a -> Decoder (NonEmpty Array a)
+decodeNonEmpty_Array decoder =
+  lmap ("Couldn't decode NonEmpty Array: " <> _)
+    <<< (traverse decoder <=< (lmap ("JSON Array" <> _) <<< rmap (\x -> x.head :| x.tail) <<< note " is empty" <<< Arr.uncons) <=< decodeJArray)
+
+decodeNonEmptyArray :: ∀ a . Decoder a -> Decoder (NonEmptyArray a)
+decodeNonEmptyArray decoder =
+  lmap ("Couldn't decode NonEmptyArray: " <> _)
+    <<< (traverse decoder <=< (lmap ("JSON Array" <> _) <<< rmap (\x -> NEA.cons' x.head x.tail) <<< note " is empty" <<< Arr.uncons) <=< decodeJArray)
+
+decodeNonEmpty_List :: ∀ a . Decoder a -> Decoder (NonEmpty List a)
+decodeNonEmpty_List decoder =
```

Comment:
Is it possible to use a single `decodeNonEmpty` which can be used with `NonEmpty List` as well as `NonEmpty Array`, in which case we can support a `decodeNonEmpty` function (thus removing `decodeNonEmpty_Array` and `decodeNonEmpty_List` from this module).

We'd then use it to implement both the `NonEmpty List` and `NonEmpty Array` instances in the `Class` module.

Reply:
But they also have different error message